File: LeetCode/409-longest-palindrome/409-longest-palindrome.js
/**
 * @param {string} s
 * @return {number}
 */
 var longestPalindrome = function (s) {
    let stringObj = {};
    for (const iterator of s) {
        if (stringObj[iterator] === undefined) {
            stringObj[iterator] = 1;
        } else {
            stringObj[iterator] += 1;
        }
    }

    let r = 0;
    let q = 0;
    for (const key in stringObj) {
        r += parseInt(stringObj[key] / 2);
        if (stringObj[key] % 2 == 1) {
            q = 1
        }
    }

    return r * 2 + q
};


// var longestPalindrome = function (s) {
//     let stringObj = {};
//     let iterator;
//     for (const iterator of s) {
//         if (Object.hasOwnProperty.call(stringObj, iterator)) {
//             stringObj[iterator] += 1;
//         } else {
//             stringObj[iterator] = 1;
//         }
//     }

//     let r = 0;
//     let q = 0;
//     let key;
//     for (key in stringObj) {
//         r += parseInt(stringObj[key] / 2);
//         if (stringObj[key] % 2 == 1) {
//             q = 1
//         }
//     }

//     return r * 2 + q
// };

let input;
let answer;
let sol;

input = "abccccdd";
answer = 7;
sol = longestPalindrome(input);
console.assert(sol === answer, `${answer} Fail`);

input = "a";
answer = 1;
sol = longestPalindrome(input);
console.assert(sol === answer, `${answer} Fail`);

File: LeetCode/409-longest-palindrome/409-longest-palindrome.go
package main

import (
	"fmt"
)

func longestPalindrome(s string) int {
	stringMap := make(map[int32]int)
	var ok bool

	for _, e := range s {

		_, ok = stringMap[e]
		if ok {
			stringMap[e] += 1
		} else {
			stringMap[e] = 1
		}
	}

	r := 0
	q := 0
	for _, v := range stringMap {
		r += int(v / 2)
		if v%2 == 1 {
			q = 1
		}

	}

	return r*2 + q
}

func main() {
	var input string
	var answer int

	input = "abccccdd"
	answer = 7
	if longestPalindrome(input) != answer {
		fmt.Printf("%v != %v fail\n", input, answer)
	}

	input = "a"
	answer = 1
	if longestPalindrome(input) != answer {
		fmt.Printf("%v != %v fail\n", input, answer)
	}
}
